cmds/md5sum/md5sum.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"crypto/md5"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+func getInput(fileName string) (input []byte, err error) {
+
+	if fileName != "" {
+		return ioutil.ReadFile(fileName)
+	}
+	return ioutil.ReadAll(os.Stdin)
+}
+
+func helpPrinter() {
+
+	fmt.Printf("Usage:\nmd5sum <File Name>\n")
+	flag.PrintDefaults()
+	os.Exit(0)
+}
+
+func versionPrinter() {
+	fmt.Println("md5sum utility, URoot Version.")
+	os.Exit(0)
+}
+
+func calculateMd5Sum(data []byte) string {
+	return fmt.Sprintf("%x", md5.Sum(data))
+}
+
+func main() {
+	var (
+		help    bool
+		version bool
+	)
+	cliArgs := ""
+	flag.BoolVar(&help, "help", false, "Show this help and exit")
+	flag.BoolVar(&version, "version", false, "Print Version")
+	flag.Parse()
+
+	if help {
+		helpPrinter()
+	}
+
+	if version {
+		versionPrinter()
+	}
+
+	if len(os.Args) >= 2 {
+		cliArgs = os.Args[1]
+	}
+	input, err := getInput(cliArgs)
```

Comment:
You don't have to keep all the contents of the file in memory.
There is hash.Hash interface implemented by all hash functions in go. This interface embeds io.Writer.
So your checksum function may look like:
```
func checksum(hasher hash.Hash, r io.Reader) {
	if _, err := io.Copy(hasher, r); err != nil {
		log.Fatal(err)
	}
	sum := hasher.Sum(nil)
	fmt.Println(hex.EncodeToString(sum))
}
```